{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "N2gAGhIyyiOW"
      },
      "outputs": [],
      "source": [
        "#Note: donor vector pDO23A has a point mutation in our stock, but it is not located in a crucial site.\n",
        "#The point mutation is at index 124 (T-->C) in the following part of the backbone:\n",
        "#reverse_complement_before_insert = \"ACTTTGTACAAGAAAGTTGAACGAGAAACGTAAAATGATATAAATATCAATATATTAAATTAGATTTTGCATAAAAAACAGACTACATAATACTGTAAAACACAACATATGCAGTCACTATGAACCAACTACTTAGATGGTATTAGTGACCTGTAGAATTCGAGCTCTAGAGCTGCAGGGCCCGAGCTTAAGACTGGCCGTCGTTTTACAACGTCGTGACTGGGAAAACATCCATGCTAGCGTTAACGCGAGAGTAGGGAACTGCCAGGCATCAAATAAAACGAAAGGCTCAGTCGGAAGACTGGGCCTTTCGTTTTATCTGTTGTTTGTCGGTGAACGCTCTCCTGAGTAGGACAAATCCGCCGGGAGCGGATTTGAACGTTGTGAAGCAACGGCCCGGAGGGTGGCGGGCAGGACGCCCGCCATAAACTGCCAGGCATCAAACTAAGCAGAAGGCCATCCTGACGGATGGCCTTTTTGCGTTTCTACAAACTCTTCCTGGCTAGCGGTACGCGTATTAATTGCGTTGCGCTCACTGCCCGCTTTCCAGTCGGGAAACCTGTCGTGCCAGCTGCATTAATGAATCGGCCAACGCGCGGGGAGAGGCGGTTTGCGTATTGGGCGCTCTTCCGCTTCCTCGCTCACTGACTCGCTGCGCTCGGTCGTTCGGCTGCGGCGAGCGGTATCAGCTCACTCAAAGGCGGTAATACGGTTATCCACAGAATCAGGGGATAACGCAGGAAAGAACATGTGAGCAAAAGGCCAGCAAAAGGCCAGGAACCGTAAAAAGGCCGCGTTGCTGGCGTTTTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGCTCAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTTCCCCCTGGAAGCTCCCTCGTGCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGCGCTTTCTCATAGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTGCACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTAT\""
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install Biopython\n",
        "from Bio.PDB import *\n",
        "from Bio.Seq import Seq\n",
        "import pandas as pd"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2L3g4G1Hq4dq",
        "outputId": "8956fa63-28b6-4b49-db45-60fd50cf55b2"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting Biopython\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from Biopython) (2.0.2)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m13.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: Biopython\n",
            "Successfully installed Biopython-1.86\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "designs = pd.read_excel(\"Design wet lab validation master spreadsheet.xlsx\")\n",
        "sequencing_results = pd.read_csv(\"1101309_summary.csv\")"
      ],
      "metadata": {
        "id": "0AMDUw5asFD8"
      },
      "execution_count": 168,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# #optional- see what nucleotides flank the gene sequence\n",
        "# first_att_list = []\n",
        "# second_att_list = []"
      ],
      "metadata": {
        "id": "-n3RQeBiRbk3"
      },
      "execution_count": 160,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def check_sequence(sequence, BP_insert):\n",
        "  before_insert = \"ATAAGGCGCAGCGGTCGGGCTGAACGGGGGGTTCGTGCACACAGCCCAGCTTGGAGCGAACGACCTACACCGAACTGAGATACCTACAGCGTGAGCTATGAGAAAGCGCCACGCTTCCCGAAGGGAGAAAGGCGGACAGGTATCCGGTAAGCGGCAGGGTCGGAACAGGAGAGCGCACGAGGGAGCTTCCAGGGGGAAACGCCTGGTATCTTTATAGTCCTGTCGGGTTTCGCCACCTCTGACTTGAGCGTCGATTTTTGTGATGCTCGTCAGGGGGGCGGAGCCTATGGAAAAACGCCAGCAACGCGGCCTTTTTACGGTTCCTGGCCTTTTGCTGGCCTTTTGCTCACATGTTCTTTCCTGCGTTATCCCCTGATTCTGTGGATAACCGTATTACCGCCTTTGAGTGAGCTGATACCGCTCGCCGCAGCCGAACGACCGAGCGCAGCGAGTCAGTGAGCGAGGAAGCGGAAGAGCGCCCAATACGCAAACCGCCTCTCCCCGCGCGTTGGCCGATTCATTAATGCAGCTGGCACGACAGGTTTCCCGACTGGAAAGCGGGCAGTGAGCGCAACGCAATTAATACGCGTACCGCTAGCCAGGAAGAGTTTGTAGAAACGCAAAAAGGCCATCCGTCAGGATGGCCTTCTGCTTAGTTTGATGCCTGGCAGTTTATGGCGGGCGTCCTGCCCGCCACCCTCCGGGCCGTTGCTTCACAACGTTCAAATCCGCTCCCGGCGGATTTGTCCTACTCAGGAGAGCGTTCACCGACAAACAACAGATAAAACGAAAGGCCCAGTCTTCCGACTGAGCCTTTCGTTTTATTTGATGCCTGGCAGTTCCCTACTCTCGCGTTAACGCTAGCATGGATGTTTTCCCAGTCACGACGTTGTAAAACGACGGCCAGTCTTAAGCTCGGGCCCTGCAGCTCTAGAGCTCGAATTCTACAGGTCACTAATACCATCTAAGTAGTTGGTTCATAGTGACTGCATATGTTGTGTTTTACAGTATTATGTAGTCTGTTTTTTATGCAAAATCTAATTTAATATATTGATATTTATATCATTTTACGTTTCTCGTTCAACTTTCTTGTACAAAGT\"\n",
        "  after_insert = \"CAACTTTATTATACAAAGTTGGCATTATAAAAAAGCATTGCTTATCAATTTGTTGCAACGAACAGGTCACTATCAGTCAAAATAAAATCATTATTTGGAGCTCCATGGTAGCGTTAACGCGGCCGCGATGTCAGGTGGCACTTTTCGGGGAAATGTGCGCGGAACCCCTATTTGTTTATTTTTCTAAATACATTCAAATATGTATCCGCTCATGAGACAATAACCCTGATAAATGCTTCAATAATATTGAAAAAGGAAGAGTATGAGTATTCAACATTTCCGTGTCGCCCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAAGATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTGAGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATTATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAGTACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAACCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTTTTTGCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCAAACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAACTACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCTGCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGTATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGGAGTCAGGCAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAACTGTCAGACCAAGTTTACTCATATATACTTTAGATTGATTTAAAACTTCATTTTTAATTTAAAAGGATCTAGGTGAAGATCCTTTTTGATAATCTCATGACCAAAATCCCTTAACGTGAGTTTTCGTTCCACTGAGCGTCAGACCCCGTAGAAAAGATCAAAGGATCTTCTTGAGATCCTTTTTTTCTGCGCGTAATCTGCTGCTTGCAAACAAAAAAACCACCGCTACCAGCGGTGGTTTGTTTGCCGGATCAAGAGCTACCAACTCTTTTTCCGAAGGTAACTGGCTTCAGCAGAGCGCAGATACCAAATACTGTTCTTCTAGTGTAGCCGTAGTTAGGCCACCACTTCAAGAACTCTGTAGCACCGCCTACATACCTCGCTCTGCTAATCCTGTTACCAGTGGCTGCTGCCAGTGGCGATAAGTCGTGTCTTACCGGGTTGGACTCAAGACGATAGTTACCGG\"\n",
        "  start_attb = \"GGGGACAGCTTTCTTGTACAAAGTGGT\"\n",
        "  end_attb = \"TACAACTTTATTATACAAAGTTGTCCCC\"\n",
        "  gene = BP_insert.replace(start_attb,\"\")\n",
        "  gene = gene.replace(end_attb,\"\")\n",
        "\n",
        "  result = \"\"\n",
        "  if sequence[0:1100] != before_insert or sequence[-1589:] != after_insert:\n",
        "    result += \" Backbone segments altered around insert.\"\n",
        "  if gene not in sequence[1100:-1589]:\n",
        "    result += \" Insert sequence is incorrect or missing.\"\n",
        "  elif sequence == before_insert + \"GGT\" + gene + \"TA\" + after_insert:\n",
        "    result = \" Correct.\"\n",
        "  else:\n",
        "    result = \" Unknown problem in sequence.\"\n",
        "\n",
        "#This commented version was used before observing that \"GGT\" and \"TA\" will always flank the gene post-recombination\n",
        "  # elif sequence[0:1100] == before_insert and sequence[-1589:] == after_insert and gene in sequence[1100:-1589]:\n",
        "  #   result = \" Correct.\"\n",
        "  #   gene_index = sequence.find(gene)\n",
        "  #   first_att = sequence[1100:gene_index]\n",
        "  #   second_att = sequence[gene_index + len(gene):-1589]\n",
        "  #   first_att_list.append(first_att)\n",
        "  #   second_att_list.append(second_att)\n",
        "\n",
        "  return result\n"
      ],
      "metadata": {
        "id": "Mg3MkvICNTMV"
      },
      "execution_count": 169,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for index in sequencing_results[\"Sample Name\"].index:\n",
        "  sample = sequencing_results.loc[index,\"Sample Name\"]\n",
        "  sequence = sequencing_results.loc[index,\"Consensus Sequence\"]\n",
        "  for index_designs in designs[\"BP colonies sequenced\"].index:\n",
        "    colonies = \"\"\n",
        "    if type(designs.loc[index_designs, \"BP colonies sequenced\"]) == str:\n",
        "      colonies = designs.loc[index_designs, \"BP colonies sequenced\"].split(\";\")\n",
        "    if sample in colonies:\n",
        "      if sequencing_results.loc[index,\"Sample Status\"] == \"Fail\":\n",
        "        if pd.isnull(designs.loc[index_designs, \"BP sequencing results\"]) == True:\n",
        "          designs.loc[index_designs, \"BP sequencing results\"] = sample + \":\" + \" Sequencing failed.\"\n",
        "        elif type(designs.loc[index_designs, \"BP sequencing results\"]) == str:\n",
        "          all_results = designs.loc[index_designs, \"BP sequencing results\"]\n",
        "          all_results += \" \" + sample + \":\" + \" Sequencing failed.\"\n",
        "          designs.loc[index_designs, \"BP sequencing results\"] = all_results\n",
        "        continue\n",
        "\n",
        "      BP_insert = designs.loc[index_designs, \"BP cloning sequence\"]\n",
        "      result = check_sequence(sequence, BP_insert)\n",
        "      if pd.isnull(designs.loc[index_designs, \"BP sequencing results\"]) == True:\n",
        "        designs.loc[index_designs, \"BP sequencing results\"] = sample + \":\" + result\n",
        "        # print(designs.loc[index_designs, \"BP sequencing results\"])\n",
        "      elif type(designs.loc[index_designs, \"BP sequencing results\"]) == str:\n",
        "        all_results = designs.loc[index_designs, \"BP sequencing results\"]\n",
        "        all_results += \" \" + sample + \":\" + result\n",
        "        designs.loc[index_designs, \"BP sequencing results\"] = all_results\n",
        "        # print(all_results)\n"
      ],
      "metadata": {
        "id": "PZUjyGvBsTTU"
      },
      "execution_count": 170,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "designs.to_excel(\"Designs_sequencing_results.xlsx\")"
      ],
      "metadata": {
        "id": "lMYml0nH9IWT"
      },
      "execution_count": 171,
      "outputs": []
    }
  ]
}